stm32-modules/vacuum-module/firmware/pump/pump_hardware.c

```diff
@@ -0,0 +1,255 @@
+    sConfigOC.OCFastMode = TIM_OCFAST_DISABLE;
+    if (HAL_TIM_PWM_ConfigChannel(&htim17, &sConfigOC, TIM_CHANNEL_1) != HAL_OK) {
+        Error_Handler();
+    }
+    HAL_TIM_MspPostInit(&htim17);
+}
+
+static void pump_tach_timer_init(void) {
+    TIM_MasterConfigTypeDef sMasterConfig = {0};
+    TIM_IC_InitTypeDef sConfigIC = {0};
+    TIM_ClockConfigTypeDef sClockSourceConfig = {0};
+
+    htim3.Instance = TACH_TIMER;
+    htim3.State = HAL_TIM_STATE_RESET;
+    htim3.Init.Prescaler = TACH_PRESCALER;
+    htim3.Init.CounterMode = TIM_COUNTERMODE_UP;
+    htim3.Init.Period = TACH_PERIOD;
+    htim3.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;
+    htim3.Init.AutoReloadPreload = TIM_AUTORELOAD_PRELOAD_DISABLE;
+    if (HAL_TIM_Base_Init(&htim3) != HAL_OK) Error_Handler();
+
+    sClockSourceConfig.ClockSource = TIM_CLOCKSOURCE_INTERNAL;
+    if (HAL_TIM_ConfigClockSource(&htim3, &sClockSourceConfig) != HAL_OK)
+        Error_Handler();
+
+    /* Initialize Input Capture mode */
+    if (HAL_TIM_IC_Init(&htim3) != HAL_OK) Error_Handler();
+
+    sMasterConfig.MasterOutputTrigger = TIM_TRGO_ENABLE;
+    sMasterConfig.MasterSlaveMode = TIM_MASTERSLAVEMODE_DISABLE;
+    if (HAL_TIMEx_MasterConfigSynchronization(&htim3, &sMasterConfig) != HAL_OK) {
+        Error_Handler();
+    }
+    /* Initialize TIM3 input capture channel */
+    sConfigIC.ICPolarity = TIM_INPUTCHANNELPOLARITY_RISING;
+    sConfigIC.ICSelection = TIM_ICSELECTION_DIRECTTI;
+    sConfigIC.ICPrescaler = TIM_ICPSC_DIV1;
+    sConfigIC.ICFilter = 0;
+    if (HAL_TIM_IC_ConfigChannel(&htim3, &sConfigIC, TACH_CHANNEL) != HAL_OK) {
+        Error_Handler();
+    }
+
+    // /* TIM3 interrupt Init */
+    HAL_NVIC_SetPriority(TACH_IRQ, 0, 0);
+    HAL_NVIC_EnableIRQ(TACH_IRQ);
+}
+
+void pump_pwm_gpio_init() {
+    GPIO_InitTypeDef GPIO_InitStruct = {0};
+    /* Peripheral clock enable */
+    __HAL_RCC_TIM17_CLK_ENABLE();
+    __HAL_RCC_GPIOA_CLK_ENABLE();
+
+    /**TIM17 GPIO Configuration
+    PA7     ------> TIM17_CH1
+    */
+    GPIO_InitStruct.Pin = PUMP_PWM_GPIO_Pin;
+    GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+    GPIO_InitStruct.Pull = GPIO_NOPULL;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+    GPIO_InitStruct.Alternate = GPIO_AF1_TIM17;
+    HAL_GPIO_Init(PUMP_PWM_GPIO_Port, &GPIO_InitStruct);
+}
+
+void pump_tach_gpio_init() {
+    GPIO_InitTypeDef GPIO_InitStruct = {0};
+    /* Peripheral clock enable */
+    __HAL_RCC_TIM3_CLK_ENABLE();
+    __HAL_RCC_GPIOB_CLK_ENABLE();
+
+    GPIO_InitStruct.Pin = PUMP_TACH_GPIO_Pin;
+    GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+    GPIO_InitStruct.Pull = GPIO_NOPULL;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+    GPIO_InitStruct.Alternate = GPIO_AF2_TIM3;
+    HAL_GPIO_Init(PUMP_TACH_GPIO_Port, &GPIO_InitStruct);
+}
+
+static uint32_t clamp(uint32_t val, uint32_t min, uint32_t max) {
+    if (val < min) return min;
+    if (val > max) return max;
+    return val;
+}
+
+bool hw_start_pump_motor(bool start) {
```

Comment:
It might get confusing if the way to turn off the pump motor is `hw_start_pump_motor(false)`